src/main/java/org/graylog/integrations/notifications/types/SlackEventNotification.java

```diff
@@ -0,0 +1,236 @@
+/**
+ * This file is part of Graylog.
+ *
+ * Graylog is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Graylog is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.graylog.integrations.notifications.types;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.floreysoft.jmte.Engine;
+import org.apache.directory.api.util.Strings;
+import org.graylog.events.notifications.*;
+import org.graylog.events.processor.EventDefinitionDto;
+import org.graylog.events.processor.aggregation.AggregationEventProcessorConfig;
+import org.graylog.integrations.notifications.modeldata.StreamModelData;
+import org.graylog.scheduler.JobTriggerDto;
+import org.graylog2.jackson.TypeReferences;
+import org.graylog2.notifications.Notification;
+import org.graylog2.notifications.NotificationService;
+import org.graylog2.plugin.MessageSummary;
+import org.graylog2.plugin.streams.Stream;
+import org.graylog2.plugin.system.NodeId;
+import org.graylog2.shared.bindings.providers.OkHttpClientProvider;
+import org.graylog2.streams.StreamService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static java.util.Objects.isNull;
+import static java.util.Objects.requireNonNull;
+
+public class SlackEventNotification implements EventNotification {
+
+	private static final String UNKNOWN_VALUE = "<unknown>";
+
+	public interface Factory extends EventNotification.Factory {
+		@Override
+        SlackEventNotification create();
+	}
+
+
+	private static final Logger LOG = LoggerFactory.getLogger(SlackEventNotification.class);
+
+	private final EventNotificationService notificationCallbackService;
+	private final Optional<StreamService> streamService ;
```

Comment:
as I said below, the streams are optional from a functional standpoint and hence the dependency should be `Optional`

Reply:
I feel like you may be misunderstanding the purpose of `Optional` in Java.

`Optional` was added to the language to help avoid the pain of dealing with NullPointerExceptions.  If you are writing a method that returns an object but may sometimes return null, using an `Optional` as your return type will help the people who call your method (including you) gracefully handle the cases when it returns null instead of the expected object.  `Optional` is intended for use as a return type.  The only reason to declare something in your own code as `Optional` is if it is being populated by a method that returns an `Optional`.

In this case, you seem to be using it as a way to signal that this particular dependency may not be needed.  That's not how `Optional` is supposed to be used.